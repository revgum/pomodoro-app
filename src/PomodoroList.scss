@import "styles/shared.scss";

.pomodoro-list {
  ul {
    list-style-type: none;
    margin: 0;
    padding: 0;

    li.hidden {
      display: none;
    }

    li.list-toggle:hover {
      cursor: pointer;
      text-decoration: underline;
    }

    li.more {
      font-weight: bold;
      font-size: 1.1rem;
      color: #ccc;
    }
  }
}
