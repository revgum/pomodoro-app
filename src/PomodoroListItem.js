// @flow
import * as React from "react";
import "./PomodoroListItem.css";
import * as Utils from "./utils";
import type { PauseState } from "./types";
import PauseList from "./PauseList";

// material-ui
import { withStyles } from "material-ui/styles";
import Card, { CardContent, CardHeader } from "material-ui/Card";
import Delete from "@material-ui/icons/Delete";
import IconButton from "material-ui/IconButton";
import red from "material-ui/colors/red";

type Props = {
  classes: any,
  deletePomodoro: Function,
  id: number,
  isPaused: boolean,
  pauses: Array<PauseState>,
  pausedElapsed: number,
  secondsRemaining: number,
  visible: boolean
};

const styles = theme => ({
  card: {
    marginBottom: theme.spacing.unit,
    padding: 0
  },
  cardContent: {
    padding: 12
  },
  deleteAction: {
    color: red[500]
  },
  pos: {
    marginBottom: 12
  }
});

class PomodoroListItem extends React.Component<Props> {
  render() {
    return (
      <li className={this.classes()}>
        <Card className={this.props.classes.card}>
          <CardHeader
            className={this.props.classes.cardHeader}
            action={
              <IconButton>
                <Delete
                  className={this.props.classes.deleteAction}
                  onClick={this.deletePomodoro}
                />
              </IconButton>
            }
            title={this.headerTitle()}
          />
          <CardContent className={this.props.classes.cardContent}>
            <PauseList
              isPaused={true}
              pauses={this.props.pauses}
              pausedElapsed={this.props.pausedElapsed}
            />
          </CardContent>
        </Card>
      </li>
    );
  }

  headerTitle = () => {
    if (this.props.secondsRemaining <= 0) return "Complete";
    return `${Utils.formattedTime(this.props.secondsRemaining)} remaining`;
  };

  classes = () => {
    return this.props.visible ? "" : "hidden";
  };

  deletePomodoro = () => {
    this.props.deletePomodoro(this.props);
  };
}

export default withStyles(styles)(PomodoroListItem);
