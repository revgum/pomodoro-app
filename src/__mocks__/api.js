import { isNullOrUndefined } from "util";

const mockInitialState = {
  test: "test"
};

export async function initialState(error) {
  if (!error) return mockInitialState;
  throw new Error("mocked error");
}

export async function save(timer_state, error) {
  if (!error) return timer_state;
  throw new Error("mocked save error");
}

export async function fetchAll(error) {
  if (!error) return [];
  throw new Error("mocked fetchAll error");
}

export async function destroy(timer_state, error) {
  if (!error) return timer_state;
  throw new Error("mocked destroy error");
}
