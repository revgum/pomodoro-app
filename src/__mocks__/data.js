export const pomodoros = [
  {
    id: 1,
    visible: true,
    classes: {},
    isPaused: false,
    isStarted: true,
    seconds: 0,
    minutes: 25,
    pauses: [],
    pausedElapsed: 0,
    secondsElapsed: 0,
    secondsRemaining: 25 * 60,
    timer: null
  }
];

export const lots_of_pomodoros = [
  pomodoros[0],
  Object.assign({}, pomodoros[0], { id: 2 }),
  Object.assign({}, pomodoros[0], { id: 3 }),
  Object.assign({}, pomodoros[0], { id: 4 })
];
