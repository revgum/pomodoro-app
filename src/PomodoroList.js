// @flow
import * as React from "react";
import "./PomodoroList.css";
import PomodoroListItem from "./PomodoroListItem";
import type { TimerState } from "./types";

// material-ui
import { withStyles } from "material-ui";
import Button from "material-ui/Button";
import Grid from "material-ui/Grid";

type Props = {
  classes: any,
  deletePomodoro: Function,
  pomodoros: Array<TimerState>
};

type State = {
  showAll: boolean
};

const styles = theme => ({
  root: {
    flexGrow: 1
  },
  button: {
    margin: theme.spacing.unit
  },
  listContainer: {
    marginTop: 0
  },
  list: {
    listStyleType: "none",
    margin: 0,
    padding: 0
  },
  listHeader: {
    color: "#ccc",
    marginBottom: theme.spacing.unit
  }
});

class PomodoroList extends React.Component<Props, State> {
  static defaultProps = {
    pomodoros: []
  };

  state = {
    showAll: false
  };

  render() {
    return (
      <div className={this.props.classes.listContainer + " pomodoro-list"}>
        <ul className={this.props.classes.list}>
          <li key="toggle" onClick={this.clickShowAll}>
            <Grid
              container
              className={this.props.classes.root}
              alignItems="flex-end"
            >
              <Grid item>
                <h2 className={this.props.classes.listHeader}>Pomodoros</h2>
              </Grid>
              <Grid item>
                {this.props.pomodoros.length > 3 && (
                  <Button
                    color={this.state.showAll ? "default" : "primary"}
                    className={this.props.classes.button}
                    size="small"
                    variant="raised"
                  >
                    {this.state.showAll ? "Less" : "View More"}
                  </Button>
                )}
              </Grid>
            </Grid>
          </li>
          {this.props.pomodoros.map((p, i) => (
            <PomodoroListItem
              key={`pomodoro-${i}`}
              {...p}
              visible={i <= 2 || this.state.showAll}
              deletePomodoro={this.props.deletePomodoro}
            />
          ))}
          {this.moreListItem()}
        </ul>
      </div>
    );
  }

  clickShowAll = () => {
    this.setState({
      showAll: !this.state.showAll
    });
  };

  moreListItem = () => {
    if (this.props.pomodoros.length <= 3) return null;
    return (
      <li key="more" className={`more ${this.state.showAll ? "hidden" : ""}`}>
        {`... ${this.props.pomodoros.length - 3} more`}
      </li>
    );
  };
}

export default withStyles(styles)(PomodoroList);
