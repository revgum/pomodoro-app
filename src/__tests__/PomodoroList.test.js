import React from "react";
import PomodoroList from "../PomodoroList";
import { createShallow, createRender } from "material-ui/test-utils";
import renderer from "react-test-renderer";
import PomodoroListItem from "../PomodoroListItem";
import * as mockData from "../__mocks__/data";

describe("<PomodoroList />", () => {
  let shallow;
  let wrapper;
  beforeAll(() => {
    shallow = createShallow();
    wrapper = shallow(<PomodoroList />).dive();
  });

  it("passes a smoke test", () => {
    const wrapper = shallow(<PomodoroList />);
  });

  it("captures a snapshot", () => {
    const tree = renderer.create(<PomodoroList />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("toggles showing all pomodoros", () => {
    const instance = wrapper.instance();
    const clickShowAllSpy = jest.spyOn(instance, "clickShowAll");
    expect(wrapper).toHaveState({
      showAll: false
    });
    instance.clickShowAll();
    expect(clickShowAllSpy).toHaveBeenCalled();
    expect(wrapper).toHaveState({
      showAll: true
    });
  });

  describe("with pomodoros", () => {
    let shallow;
    let wrapper;
    let instance;
    beforeAll(() => {
      shallow = createShallow();
      wrapper = shallow(<PomodoroList pomodoros={mockData.pomodoros} />).dive();
      instance = wrapper.instance();
    });

    it("renders a pomodoro list item", () => {
      expect(wrapper.find(PomodoroListItem)).toExist();
    });
    it("hides the ...more list item with fewer than 4 pomodoros", () => {
      const moreListItemSpy = jest.spyOn(instance, "moreListItem");
      expect(wrapper.find(".more")).not.toExist();
      let list_item = instance.moreListItem();
      expect(moreListItemSpy).toHaveBeenCalled();
      expect(list_item).toBeNull();
    });
    it("shows the ...more list item with more than 3 pomodoros", () => {
      const moreListItemSpy = jest.spyOn(instance, "moreListItem");
      let new_wrapper = shallow(
        <PomodoroList
          classes={{}}
          deletePomodoro={jest.fn()}
          pomodoros={mockData.lots_of_pomodoros}
        />
      ).dive();
      expect(new_wrapper.find("li.more.hidden")).not.toExist();
      expect(new_wrapper.find("li.more")).toExist();
    });
  });
});
