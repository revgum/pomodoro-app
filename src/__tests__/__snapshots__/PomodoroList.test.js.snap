// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PomodoroList /> captures a snapshot 1`] = `
<div
  className="PomodoroList-listContainer-3 pomodoro-list"
>
  <ul
    className="PomodoroList-list-4"
  >
    <li
      onClick={[Function]}
    >
      <div
        className="MuiGrid-container-6 MuiGrid-align-items-xs-flex-end-16 PomodoroList-root-1"
      >
        <div
          className="MuiGrid-item-7"
        >
          <h2
            className="PomodoroList-listHeader-5"
          >
            Pomodoros
          </h2>
        </div>
        <div
          className="MuiGrid-item-7"
        />
      </div>
    </li>
  </ul>
</div>
`;
